rosetta/transactor/transactor.go

```diff
@@ -0,0 +1,592 @@
+	}
+
+	// Create the signers list.
+	signers := []identifier.Account{
+		sender,
+	}
+
+	return rosettaBlockID(height, blockID), tx.ProposalKey.SequenceNumber, ops, signers, nil
+}
+
+// AttachSignature adds the given signature to the transaction.
+func (t *Transactor) AttachSignature(unsigned string, signature object.Signature) (string, error) {
+
+	unsignedTx, err := t.decodeTransaction(unsigned)
+	if err != nil {
+		return "", fmt.Errorf("could not decode transaction: %w", err)
+	}
+
+	// Validate the transaction actors. We expect a single authorizer - the sender account.
+	if len(unsignedTx.Authorizers) != authorizersRequired {
+		return "", failure.InvalidAuthorizers{
+			Have:        uint(len(unsignedTx.Authorizers)),
+			Want:        authorizersRequired,
+			Description: failure.NewDescription("invalid number of authorizers"),
+		}
+	}
+
+	sender := unsignedTx.Authorizers[0]
+
+	// Verify that the sender is the payer, since it is the payer that needs to sign the envelope.
+	if unsignedTx.Payer != sender {
+		return "", failure.InvalidPayer{
+			Have:        flow.BytesToAddress(unsignedTx.Payer[:]),
+			Want:        flow.BytesToAddress(sender[:]),
+			Description: failure.NewDescription("invalid transaction payer"),
+		}
+	}
+
+	// Verify that we do not already have signatures.
+	if len(unsignedTx.EnvelopeSignatures) > 0 {
+		return "", failure.InvalidSignature{
+			Description: failure.NewDescription("unexpected envelope signatures found",
+				failure.WithInt("signatures", len(unsignedTx.EnvelopeSignatures))),
+		}
+	}
+
+	// Verify that the signature belongs to the sender.
+	signer := sdk.HexToAddress(signature.SigningPayload.AccountID.Address)
+	if signer != sender {
+		return "", failure.InvalidSignature{
+			Description: failure.NewDescription("invalid signer account",
+				failure.WithString("have_signer", signer.String()),
+				failure.WithString("want_signer", sender.String())),
+		}
+	}
+
+	bytes, err := hex.DecodeString(signature.HexBytes)
+	if err != nil {
+		return "", failure.InvalidSignature{
+			Description: failure.NewDescription("invalid signature payload",
+				failure.WithErr(err)),
+		}
+	}
+
+	// TODO: allow arbitrary key index
+	// => https://github.com/optakt/flow-dps/issues/369
+	signedTx := unsignedTx.AddEnvelopeSignature(signer, 0, bytes)
+	signed, err := t.encodeTransaction(signedTx)
+	if err != nil {
+		return "", fmt.Errorf("could not encode transaction: %w", err)
+	}
+
+	return signed, nil
+}
+
+func (t *Transactor) verifySignature(signed string, address flow.Address) error {
+
+	signedTx, err := t.decodeTransaction(signed)
+	if err != nil {
+		return fmt.Errorf("could not decode transaction: %w", err)
+	}
+
+	if len(signedTx.EnvelopeSignatures) != 1 {
+		return failure.InvalidSignature{
+			Description: failure.NewDescription("invalid number of signatures",
+				failure.WithInt("signatures", len(signedTx.EnvelopeSignatures))),
+		}
+	}
+
+	rosBlockID := identifier.Block{Hash: signedTx.ReferenceBlockID.Hex()}
```

Comment:
As mentioned before, I think this should be done on whatever the latest block is, not the reference block.

Reply:
I don't think so. I think it's up to the consumer of the API to make sure the keys remain valid between the block when the transaction is created and when it is submitted. It's fine to just have a failing transaction on submission if they don't.